File: NCAA/scripts/_sweep_thresh.mjs
#!/usr/bin/env node
import fs from "fs";

const store = JSON.parse(fs.readFileSync("data/history.json", "utf8"));
const runs = store.runs || [];

const games = [];
for (const r of runs) {
  for (const g of r.games) {
    if (g.homeScore === undefined || g.awayScore === undefined) continue;
    if (!g.line && g.line !== 0) continue;
    if (g.status === "SKIPPED" || g.status === "MISSING_STATS") continue;
    if (!g.pHomeCover && !g.pAwayCover) continue;
    games.push(g);
  }
}

const totalDays = new Set(runs.map(r => r.date)).size;

// LINE CONVENTION: line > 0 = home FAV, line < 0 = away FAV
// So: home is DOG when line < 0, away is DOG when line > 0

function gradeSpread(g, side) {
  // margin_mean = hS - aS - line => home covers when margin + line > ...
  // Actually: home covers spread when homeScore - awayScore > line (for fav)
  // or homeScore - awayScore + |line| > 0 (for dog getting points)
  // With convention line > 0 = home fav:
  //   Home pick covers: (homeScore - awayScore) - line > 0 (if fav, needs to win by more than line)
  //   Wait, need to think about this more carefully.
  //
  // Standard: if home is -7 fav (our line=7), home covers if they win by > 7
  // homeMargin = homeScore - awayScore
  // home covers: homeMargin > line (when line > 0, home fav)
  // home covers: homeMargin - line > 0
  //
  // If home is +7 dog (our line=-7), home covers if they lose by < 7 or win
  // home covers: homeMargin > line => homeMargin > -7 => homeMargin + 7 > 0
  // Same formula: homeMargin - line > 0

  const homeMargin = g.homeScore - g.awayScore;
  const v = homeMargin - g.line; // positive = home covers

  if (side === "home") {
    return v > 0 ? "WIN" : v === 0 ? "PUSH" : "LOSS";
  } else {
    return v < 0 ? "WIN" : v === 0 ? "PUSH" : "LOSS";
  }
}

// SANITY CHECK: all dogs should be ~50%
let dogW = 0, dogL = 0;
for (const g of games) {
  if (g.line === 0) continue;
  // home is dog when line < 0, away is dog when line > 0
  const dogSide = g.line < 0 ? "home" : "away";
  const r = gradeSpread(g, dogSide);
  if (r === "WIN") dogW++;
  else if (r === "LOSS") dogL++;
}
console.log(`SANITY: ALL dogs ATS = ${dogW}-${dogL} (${(dogW/(dogW+dogL)*100).toFixed(1)}%) — should be ~50%`);
console.log(`Total games: ${games.length}, Total days: ${totalDays}\n`);

// === ALL SIDES ===
console.log("=== ALL SIDES (best pCover side, fav line cap 6) ===");
console.log("pCover | sDiffMin | W    - L   - P | Pct%  | Picks/Day | Units");
console.log("-------+----------+----------------+-------+-----------+------");

for (const pThresh of [0.55, 0.58, 0.60, 0.62, 0.63, 0.65, 0.67, 0.70]) {
  for (const sdMin of [0, 3]) {
    let w = 0, l = 0, p = 0;
    for (const g of games) {
      const pH = g.pHomeCover || 0;
      const pA = g.pAwayCover || 0;
      const bestP = Math.max(pH, pA);
      const side = pH >= pA ? "home" : "away";
      const absLine = Math.abs(g.line);
      const sd = g.sDiff || 0;

      // Is picked side the dog?
      // home is dog when line < 0, away is dog when line > 0
      const isDog = side === "home" ? g.line < 0 : g.line > 0;
      const lineOK = isDog ? true : absLine <= 6;

      if (bestP >= pThresh && sd >= sdMin && sd <= 12 && lineOK && absLine > 0) {
        const result = gradeSpread(g, side);
        if (result === "WIN") w++;
        else if (result === "LOSS") l++;
        else p++;
      }
    }
    const total = w + l;
    const pct = total > 0 ? (w / total * 100).toFixed(1) : "n/a";
    const units = (w * 1 - l * 1.1).toFixed(1);
    const ppd = ((w + l + p) / totalDays).toFixed(1);
    console.log(`  ${pThresh.toFixed(2)}  |    ${sdMin}     | ${String(w).padStart(4)}-${String(l).padStart(4)}-${String(p).padStart(2)} | ${pct.padStart(5)} | ${ppd.padStart(9)} | ${units.padStart(6)}`);
  }
}

// === DOGS ONLY ===
console.log("\n=== DOGS ONLY (dog's pCover directly) ===");
console.log("pCover | minLine | W    - L   - P | Pct%  | Picks/Day | Units");
console.log("-------+---------+----------------+-------+-----------+------");

for (const pThresh of [0.50, 0.52, 0.53, 0.55, 0.57, 0.58, 0.60, 0.62, 0.65]) {
  for (const minLine of [0.5, 2.5, 3.5]) {
    let w = 0, l = 0, p = 0;
    for (const g of games) {
      const absLine = Math.abs(g.line);
      if (absLine < minLine) continue;

      // home is dog when line < 0, away is dog when line > 0
      const homeDog = g.line < 0;
      const awayDog = g.line > 0;
      if (!homeDog && !awayDog) continue;

      const dogSide = homeDog ? "home" : "away";
      const dogPCover = homeDog ? (g.pHomeCover || 0) : (g.pAwayCover || 0);

      if (dogPCover >= pThresh) {
        const result = gradeSpread(g, dogSide);
        if (result === "WIN") w++;
        else if (result === "LOSS") l++;
        else p++;
      }
    }
    const total = w + l;
    const pct = total > 0 ? (w / total * 100).toFixed(1) : "n/a";
    const units = (w * 1 - l * 1.1).toFixed(1);
    const ppd = ((w + l + p) / totalDays).toFixed(1);
    console.log(`  ${pThresh.toFixed(2)}  |  ${minLine.toFixed(1).padStart(4)} | ${String(w).padStart(4)}-${String(l).padStart(4)}-${String(p).padStart(2)} | ${pct.padStart(5)} | ${ppd.padStart(9)} | ${units.padStart(6)}`);
  }
}

File: pyNCAA/scripts/_sweep_thresh.py
#!/usr/bin/env python3
# scripts/_sweep_thresh.py — Full threshold sweep (spreads)
import json

store = json.loads(open("data/history.json").read())
runs = store.get("runs") or []
games = []
for r in runs:
    for g in (r.get("games") or []):
        if g.get("homeScore") is None or g.get("awayScore") is None: continue
        if g.get("line") is None and g.get("line") != 0: continue
        if g.get("status") in ("SKIPPED", "MISSING_STATS"): continue
        if not g.get("pHomeCover") and not g.get("pAwayCover"): continue
        games.append(g)

total_days = len(set(r.get("date") for r in runs))

def grade_spread(g, side):
    margin = g["homeScore"] - g["awayScore"]
    v = margin + g["line"]
    if side == "home": return "WIN" if v > 0 else ("PUSH" if v == 0 else "LOSS")
    return "WIN" if v < 0 else ("PUSH" if v == 0 else "LOSS")

# Sanity check
dog_w, dog_l = 0, 0
for g in games:
    if g.get("line", 0) == 0: continue
    dog_side = "home" if g["line"] > 0 else "away"
    r = grade_spread(g, dog_side)
    if r == "WIN": dog_w += 1
    elif r == "LOSS": dog_l += 1
print(f"SANITY: ALL dogs ATS = {dog_w}-{dog_l} ({dog_w / (dog_w + dog_l) * 100:.1f}%) -- should be ~50%")
print(f"Total games: {len(games)}, Total days: {total_days}\n")

print("=== ALL SIDES (best pCover side, fav line cap 6) ===")
print("pCover | sDiffMin | W    - L   - P | Pct%  | Picks/Day | Units")
for p_thresh in [0.55, 0.58, 0.60, 0.62, 0.63, 0.65, 0.67, 0.70]:
    for sd_min in [0, 3]:
        w, l, p = 0, 0, 0
        for g in games:
            pH, pA = g.get("pHomeCover", 0), g.get("pAwayCover", 0)
            best_p = max(pH, pA)
            side = "home" if pH >= pA else "away"
            abs_line = abs(g.get("line", 0))
            sd = g.get("sDiff", 0)
            is_dog = (side == "home" and g["line"] > 0) or (side == "away" and g["line"] < 0)
            line_ok = True if is_dog else abs_line <= 6
            if best_p >= p_thresh and sd >= sd_min and sd <= 12 and line_ok and abs_line > 0:
                result = grade_spread(g, side)
                if result == "WIN": w += 1
                elif result == "LOSS": l += 1
                else: p += 1
        total = w + l
        pct = f"{w / total * 100:.1f}" if total > 0 else "n/a"
        units = f"{(w - l * 1.1):.1f}"
        ppd = f"{(w + l + p) / total_days:.1f}"
        print(f"  {p_thresh:.2f}  |    {sd_min}     | {w:>4}-{l:>4}-{p:>2} | {pct:>5} | {ppd:>9} | {units:>6}")

print("\n=== DOGS ONLY (dog's pCover directly) ===")
print("pCover | minLine | W    - L   - P | Pct%  | Picks/Day | Units")
for p_thresh in [0.50, 0.52, 0.53, 0.55, 0.57, 0.58, 0.60, 0.62, 0.65]:
    for min_line in [0.5, 2.5, 3.5]:
        w, l, p = 0, 0, 0
        for g in games:
            abs_line = abs(g.get("line", 0))
            if abs_line < min_line: continue
            home_dog = g.get("line", 0) > 0
            away_dog = g.get("line", 0) < 0
            if not home_dog and not away_dog: continue
            dog_side = "home" if home_dog else "away"
            dog_p_cover = g.get("pHomeCover", 0) if home_dog else g.get("pAwayCover", 0)
            if dog_p_cover >= p_thresh:
                result = grade_spread(g, dog_side)
                if result == "WIN": w += 1
                elif result == "LOSS": l += 1
                else: p += 1
        total = w + l
        pct = f"{w / total * 100:.1f}" if total > 0 else "n/a"
        units = f"{(w - l * 1.1):.1f}"
        ppd = f"{(w + l + p) / total_days:.1f}"
        print(f"  {p_thresh:.2f}  |  {min_line:4.1f} | {w:>4}-{l:>4}-{p:>2} | {pct:>5} | {ppd:>9} | {units:>6}")
